### DISCUSSION TITLE
Don't share data between different tabbed windows.

### DISCUSSION THREAD

**Discussion by coderjXD** [pointfreeco/swift-composable-architecture#2883] [Q&A] [answered]
I am composing a macOS app which can be multiple tabbed(Not the TabView, but tabs that you can create with the ⌘+T shortcut). How can I make these different tabbed windows not share the same store? With the general feature implementation, any change in one tab/window will reflect to all other tabs and windows. How should the feature be architectured for my purpose?

**Comment by oliverfoggin** [comment]
Hmm... I'm not sure that your request specifically follows from your requirement.

You can definitely have a single store with multiple tabs and NOT have them sharing data.

Can you explain a bit more about your requirements and what issues you have seen so that we can help a bit better?

Thanks

**Reply by coderjXD** [reply]
<img width="496" alt="Screenshot 2024-03-05 at 10 22 02" src="https://github.com/pointfreeco/swift-composable-architecture/assets/97173695/8f6c792b-0161-45f6-81a2-ba927de045ac">

By saying `tab` I mean the window that comes with the `＋` in this image on the system tool bar,  which you can also arouse by shortcut `⌘+T` in many apps like Xcode or Chrome. It is NOT a `TabView` in SwiftUI, but rather a window instance in the root window group. 
`
let store = Store(...)
var body: some Scene {
    WindowGroup {
        AppContent(store: store)
    }
}
`
This is the code that I currently use, which leads to tabbed windows sharing the same store. I have tried to not pass the store in from outside, like this:
`
var body: some Scene {
    WindowGroup {
        AppContent()
    }
}

struct AppContent: View {
    let store = Store(...)
}
`
But every binding in the AppContent seems to be broken because the store is no longer bindable and I can not use the `binding` method of the store or the `sending` method of its states to generate a binding.

I believe by not passing a shared store into the AppContent my requirement could be met. But I don't know how to form the bindings in the `AppContent` view. Any ideas?

**Reply by mbrandonw** [reply]
Hi @coderjXD, this is definitely your problem:

```swift
let store = Store(...) var body: some Scene { WindowGroup { AppContent(store: store) } } 
```

Every window is getting the exact same `Store` and so it shouldn't be surprising that they all show the same data. One approach is to generate a new `Store` for each window, or you can keep a collection of states in your feature that represents each window.

**Reply by coderjXD** [reply]
Yes, I believe so! I think let the store live only in the `AppContent ` will address the sharing data issue as this:
```swift
var body: some Scene {
WindowGroup {
AppContent()
}
}

struct AppContent: View {
let store = Store(...)
}
```
But since I changed the store in the `AppContent` from a `@Bindable var` to a `let`, all the previous bindings are broken. Now the issue comes to `How to build a binding with a plain store rather than a bindable store?` . With a plain store, the `binding` method of the store, and the `sending` method of its states are unavailable.

Thanks so much for your answering XD

**Reply by mbrandonw** [reply]
@coderjXD Is there a reason you changed `@Bindable var` to `let`? You can keep it `@Bindable` store.

**Reply by coderjXD** [reply, marked as answer]
Because when I use `@Bindable var`, the store is still shared between tabbed windows.
Since I realize that this is not a tca related issue, I'll close this topic. Thanks for your time XD.

**Comment by kamcma** [comment]
I don't know if you're in the Slack, but [this thread](https://pointfreecommunity.slack.com/archives/C062TFT9M54/p1709230093264369) shows upcoming `@Shared` state feature will likely make multi-window support better.

That said, this is not a TCA specific issue. I'm not a Mac dev myself, so I'm not an expert, but there are certain ways you need to use the `WindowGroup` APIs to either make all window instances have the same or independent state.

**Reply by coderjXD** [reply]
No, I'm not in the Slack. Do I have to be a subscriber of pointfree.co to be invited? If not, can you invite me in? Thanks a lot @mbrandonw . Btw, I'd like to subscribe, but my debit card is not supported 😌. I am in the mainland of China.

**Reply by stephencelis** [reply]
@coderjXD You can join the Slack here: https://www.pointfree.co/slack-invite
